soc/xtensa/intel_adsp/common/soc.c

```diff
@@ -195,34 +196,74 @@ static inline void soc_set_power_and_clock(void)
 	volatile struct soc_dsp_shim_regs *dsp_shim_regs =
 		(volatile struct soc_dsp_shim_regs *)SOC_DSP_SHIM_REG_BASE;
 
+#ifdef CONFIG_SOC_SERIES_INTEL_CAVS_V15
 	/*
+	 * HP domain clocked by PLL
+	 * LP domain clocked by PLL
 	 * DSP Core 0 PLL Clock Select divide by 1
 	 * DSP Core 1 PLL Clock Select divide by 1
-	 * Low Power Domain Clock Select depends on LMPCS bit
-	 * High Power Domain Clock Select depands on HMPCS bit
-	 * Low Power Domain PLL Clock Select device by 4
 	 * High Power Domain PLL Clock Select device by 2
-	 * Tensilica Core Prevent Audio PLL Shutdown (TCPAPLLS)
-	 * Tensilica Core Prevent Local Clock Gating (Core 0)
-	 * Tensilica Core Prevent Local Clock Gating (Core 1)
+	 * Low Power Domain PLL Clock Select device by 4
+	 * Disable Tensilica Core Prevent Audio PLL Shutdown (TCPAPLLS)
+	 * Disable Tensilica Core Prevent Local Clock Gating (Core 0)
+	 * Disable Tensilica Core Prevent Local Clock Gating (Core 1)
+	 *   - Disabling "prevent clock gating" means allowing clock gating
 	 */
 	dsp_shim_regs->clkctl =
-		SOC_CLKCTL_DPCS_DIV1(0) |
-		SOC_CLKCTL_DPCS_DIV1(1) |
-		SOC_CLKCTL_LDCS_LMPCS |
-		SOC_CLKCTL_HDCS_HMPCS |
-		SOC_CLKCTL_LPMEM_PLL_CLK_SEL_DIV4 |
-		SOC_CLKCTL_HPMEM_PLL_CLK_SEL_DIV2 |
-		SOC_CLKCTL_TCPAPLLS |
-		SOC_CLKCTL_TCPLCG_DIS(0) |
-		SOC_CLKCTL_TCPLCG_DIS(1);
-
-	/* Disable power gating for both cores */
-	dsp_shim_regs->pwrctl |= SOC_PWRCTL_DISABLE_PWR_GATING_DSP1 |
-		SOC_PWRCTL_DISABLE_PWR_GATING_DSP0;
+		SHIM_CLKCTL_HDCS_PLL |
+		SHIM_CLKCTL_LDCS_PLL |
+		SHIM_CLKCTL_DPCS_DIV1(0) |
+		SHIM_CLKCTL_DPCS_DIV1(1) |
+		SHIM_CLKCTL_HPMPCS_DIV2 |
+		SHIM_CLKCTL_LPMPCS_DIV4 |
+		SHIM_CLKCTL_TCPAPLLS_DIS |
+		SHIM_CLKCTL_TCPLCG_DIS(0) |
+		SHIM_CLKCTL_TCPLCG_DIS(1);
 
 	/* Rewrite the low power sequencing control bits */
 	dsp_shim_regs->lpsctl = dsp_shim_regs->lpsctl;
+#endif /* CONFIG_SOC_SERIES_INTEL_CAVS_V15 */
+
+#if defined(CONFIG_SOC_SERIES_INTEL_CAVS_V18) || \
+	defined(CONFIG_SOC_SERIES_INTEL_CAVS_V20) || \
+	defined(CONFIG_SOC_SERIES_INTEL_CAVS_V25)
+
+	/*
+	 * Request HP ring oscillator and
+	 * wait for status to indicate it's ready.
+	 */
+	dsp_shim_regs->clkctl |= SHIM_CLKCTL_RHROSCC;
+	while ((dsp_shim_regs->clkctl & SHIM_CLKCTL_RHROSCC) !=
+	       SHIM_CLKCTL_RHROSCC) {
+		k_busy_wait(10);
```

Comment:
We should have a timeout here as this could spin forever if HP ring fails to start (e.g. host misconfiguration of IP)

Reply:
The SOF code spins forever too. Since this is SoC init code, there is no recovery path. Do you still want me to add some retry limits? I could make this into fatal error if all retries are exhausted.